### PR TITLE
Support for NxtPX4v2

### PR DESCRIPTION
**Pull request by Minderring**
The NxtPX4v2 is an open-source hardware designed and maintened by [HKUST UAV-Group](https://github.com/HKUST-Aerial-Robotics/Nxt-FC). And it is produced by [MicoAir Tech.](https://micoair.com/).

Waiting for your reply!
Best Regard!

### CODE REVIEW

File: Tools/AP_Bootloader/board_types.txt [hunk 1]
```diff
@@ -266,6 +266,7 @@ AP_HW_MicoAir405v2                   1150
 AP_HW_ORAQF405PRO                    1155
 AP_HW_CBU_StampH743                  1156
 AP_HW_FOXEERF405_V2                  1157
+AP_HW_NxtPX4v2                       1159
```
Comment (peterbarker) [inline Tools/AP_Bootloader/board_types.txt:269]: I suggest creating a PR to reserve an ID in case this PR takes a while to get in.

Comment (Minderring) [reply Tools/AP_Bootloader/board_types.txt:269]: Thank you for your suggestion. I have cerated the PR: [#26746](https://github.com/ArduPilot/ardupilot/pull/26746)

File: libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/defaults.parm [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#BATTERY MONITOR ON
+BATT_MONITOR 4
+
+#Serial Port defaults
+SERIAL4_PROTOCOL 42
+SERIAL6_PROTOCOL 23
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/defaults.parm:6]: You should add protocol 16 for SERIAL5

**Review by andyp1per — APPROVED** [review]
LGTM but I think you need to update the board id and maybe set the ESC telemetry protocol

**Comment by peterbarker** [comment]
@Minderring needs a rebase

File: libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/defaults.parm (hunk 2 above)
Comment (Minderring) [reply libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/defaults.parm:7]: Okay! fixed.

**Comment by Minderring** [comment]
The board ID has been merged and rebased. @peterbarker @andyp1per @Hwurzburg 
If there is anything that needs to be modified, please let me know.

**Comment by peterbarker** [comment]
> The board ID has been merged and rebased. @peterbarker @andyp1per @Hwurzburg If there is anything that needs to be modified, please let me know.

I've marked it as MergeOnCIPass

**Comment by Hwurzburg** [comment]
I need to review..will do so ASAP

File: libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/hwdef.dat [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+PD15 LED_RED   OUTPUT HIGH GPIO(90)
+PD11 LED_GREEN OUTPUT HIGH GPIO(91)
+PB15 LED_BLUE  OUTPUT HIGH GPIO(92)
+
+define HAL_GPIO_A_LED_PIN 90
+define HAL_GPIO_B_LED_PIN 91
+define HAL_GPIO_C_LED_PIN 92
+
+define HAL_GPIO_LED_ON 1
+define HAL_GPIO_LED_OFF 0
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/hwdef.dat:103]: remove this line...board will not build with this line!

File: libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
+# NxtPX4v2 Flight Controller
+
+The NxtPX4v2 is an open-source hardware designed and maintened by [HKUST UAV-Group](https://github.com/HKUST-Aerial-Robotics/Nxt-FC). And it is produced by [MicoAir Tech.](http://micoair.com/).
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md:3]: if open source should submit schematics,etc. to the ArduPilot/Schematics repo

File: libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+- 7x UART
+- 8x PWM
+- 1x I2C
+- 1x SPI
+- 1x SWD
+
+## UART Mapping
+
+ - SERIAL0 -> USB
+ - SERIAL1 -> UART2 (DMA-enabled)
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md:21]:
```suggestion
 - SERIAL1 -> UART2 (MAVLink2,DMA-enabled)
```

File: libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+ - SERIAL2 -> UART4 (DMA-enabled)
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md:22]:
```suggestion
 - SERIAL2 -> UART4 (MAVLink2, DMA-enabled)
```

File: libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+ - SERIAL3 -> UART1 (GPS, DMA-enabled)
+ - SERIAL4 -> UART3 (VTX-HD, DMA-enabled)
+ - SERIAL5 -> UART7 (ESC Telemetry, DMA-enabled)
+ - SERIAL6 -> UART5 (RCIN, DMA-enabled)
+ - SERIAL7 -> UART8 (DMA-enabled)
+
+## RC Input
+
+The default RC input is configured on the UART5_RX inverted from the SBUS pin. Other RC  protocols  should be applied at other UART port such as UART2, UART4 or UART8, and set the protocol to receive RC data: `SERIALn_PROTOCOL=23` and change SERIAL6 _Protocol to something other than '23'
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md:31]:
```suggestion
The default RC input is configured on the UART5. The SBUS pin is inverted and connected to RX5. Non SBUS,  single wire serial inputs can be directly tied to RX5 if SBUS pin is left unconnected. RC could  be applied instead at a different UART port such as UART2, UART4 or UART8, and set the protocol to receive RC data: `SERIALn_PROTOCOL=23` and change SERIAL6 _Protocol to something other than '23'
```
this assumes that the SBUS inverter does not interfere with RX5 if unconnected. This is the normal configuration for most boards. If not, I can propose a different wording.

File: libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+## VTX Support
+
+The SH1.0-6P connector supports a DJI O3 Air Unit connection. Pin 1 of the connector is 12v so be careful not to connect this to a peripheral requiring 5v.
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md:35]:
```suggestion
The SH1.0-6P connector supports a DJI Air Unit / HD VTX connection. Protocol defaults to DisplayPort. Pin 1 of the connector is 12v so be careful not to connect this to a peripheral requiring 5v.
```

**Review by Hwurzburg — CHANGES_REQUESTED** [review]
a few tweaks and one major HWDEF issue that will prevent it being built
also the back view image needs to have TX4/RX4 labels instead of just TX/RX on the SERIAL2 connector

File: libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/hwdef.dat (hunk 3 above)
Comment (Minderring) [reply libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/hwdef.dat]: Okay. I have remove the line: define HAL_GPIO_LED_OFF 0

File: libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md (hunk 4 above)
Comment (Minderring) [reply libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md:3]: Got it. I have submitted the PCB files to the ArduPilot/Schematics repo. [#10](https://github.com/ArduPilot/Schematics/pull/10)
Please take a look.

File: libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md (hunk 5 above)
Comment (Minderring) [reply libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md]: Fixed.

File: libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md (hunk 6 above)
Comment (Minderring) [reply libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md]: Fixed.

File: libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md (hunk 7 above)
Comment (Minderring) [reply libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md]: Fixed. Thanks!

File: libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md (hunk 8 above)
Comment (Minderring) [reply libraries/AP_HAL_ChibiOS/hwdef/NxtPX4v2/README.md]: Fixed.

**Comment by Minderring** [comment]
> a few tweaks and one major HWDEF issue that will prevent it being built also the back view image needs to have TX4/RX4 labels instead of just TX/RX on the SERIAL2 connector

Thank you for your suggestion! I have fixed them all. Please take a look. @Hwurzburg

**Review by Hwurzburg — APPROVED** [review]
LGTM

**Comment by Minderring** [comment]
@peterbarker Hi! I have rebased and please take a look.

**Comment by Hwurzburg** [comment]
@Minderring again, please describe the changes in the the force pushes since reviews were done...otherwise we will have to wait until we can find time to do a complete re-review

**Comment by Minderring** [comment]
> @Minderring again, please describe the changes in the the force pushes since reviews were done...otherwise we will have to wait until we can find time to do a complete re-review

Sorry! Just a minor change since the last approved review.
 I corrected some errors in the back_view image regarding the AUX label, then I rebased it, and that was all.

**Comment by Hwurzburg** [comment]
lets us know when testing is completed and will mark for DevCall and merge

**Comment by Minderring** [comment]
> lets us know when testing is completed and will mark for DevCall and merge

@Hwurzburg Thanks! It was thoroughly tested last month and continues to perform well.

**Review by tridge — APPROVED** [review]

**Pull request merged**